### ISSUE TITLE
Clean up subscriptions if unsubscribe logic throws

### ISSUE THREAD

**Issue by blesh** [zenparsing/es-observable#82]
During the tear down phase of a chain of observables, one of the unsubscribe calls could throw, but the remaining unsubscription logic should still be called.

Scenario:

```js
let source = new Observable(observer => {
  setTimeout(() => {
    observer.next('hello');
    observer.complete();
  }, 1000);
  return () => {
    console.log('unsub source called');
  }
});

function brokenMap(source, project) {
  return new Observable(observer => {
    let innerSub = source.subscribe({
      next: (value) => observer.next(project(value)),
      error: observer.error.bind(observer),
      complete: observer.complete.bind(observer)
    });
    return () => {
       console.log('brokenMap unsub called');
       throw new Error('ha ha');
       innerSub.unsubscribe(); //<-- won't make it here
    }
  });
}

let mapped = brokenMap(source, x => x + '!!!');

const subscription = mapped.subscribe({
  next: console.log.bind(console)
});

setTimeout(() => subscription.unsubscribe(), 500);
```

Ideally `"unsub source called"` will be logged to console.

I haven't tested this with your reference implementation. I just wanted to put the issue up for your attention.

**Comment by blesh** [comment]
FWIW: RxJS 5 _is_ handling this scenario.

**Comment by benjamingr** [comment]
Yes, es-observable should handle it - +1

**Comment by zenparsing** [comment]
What if `observer.complete()` is not called in the example?  Do you think the source cleanup should still be run?

**Comment by zenparsing** [comment]
It looks to me like es-observable and RxJS do the same thing here.

- If `observer.complete` is called (as in the original example), then the source cleanup function runs.
- If `observer.complete` is *not* called, then the source cleanup function does not run.

Which makes sense to me.

**Comment by blesh** [comment]
> What if observer.complete() is not called in the example? Do you think the source cleanup should still be run?

Yes. 

This is more about having the consumer signal it wants to unsubscribe, then having an error thrown in the unsubscribe logic. You still want to unsubscribe from all of your other subscriptions if you're chaining them together. While es-observable's Subscription isn't really a collection of child subscriptions, I wanted to bring this up for your consideration.

**Comment by blesh** [comment]
It seems like it's really down to the user to properly implement their chained subscriptions. Making sure to call `unsubscribe` on child subscriptions *first* before doing anything volatile.

**Comment by blesh** [comment]
I've updated my example above, it was missing a bit and was poorly implemented. This probably isn't a concern for es-observable since the Subscription class is so simple.

**Comment by blesh** [comment]
TL;DR: make sure to unsub from inner subscriptions *first* in your teardown function.

**Comment by blesh** [comment]
Anything more robust would involve improving the Subscription type dramatically, which is out of scope for this spec I think.
